### DISCUSSION TITLE
How do I use int64 instead of float64 for number field?

### DISCUSSION THREAD

**Discussion by silenceofthewans** [pocketbase/pocketbase#3113] [Q&A] [answered]
I want to store currency in db and based on some research online it is generally recommended to store it as cents in int instead of float but currently the number field uses float64 and for my purpose I would need to use a custom model to store it as int64. So, is it fine to use number field as is for an app that will need to deal with currency (not actual money but just price tables for product etc). If not, then are there plans for allowing user to specify a number type for the number field from the admin dashboard or is custom model the only way for the foreseeable future?

**Comment by ganigeorgiev** [comment, marked as answer]
> So, is it fine to use number field as is for an app that will need to deal with currency (not actual money but just price tables for product etc).

I think the `number` field should suffice, but it will depend on your use case and the maximum value you intend to store. Note that because of the float64 cast, the maximum safe int value is around ~2^53-1 (aka. similar to the JS Number).

> ... are there plans for allowing user to specify a number type for the number field from the admin dashboard

I'll consider adding a `number` field toggle option to restrict int or float only numbers as validator, but the maximum safe integer that the `number` field can represent would remain as it is (big int as json representation is troublesome because not all json decoders know how to interpret them; as a safer workaround a plain `text` field could be used instead).

**Comment by stamler** [comment]
I'm coming up against this problem pretty hard now (all numbers being floats). The math is just plain wrong due to IEEE754 rounding errors, even with SUM(). Please consider letting us specify INT or better DECIMAL values (using the decimal.c extension)

**Reply by ganigeorgiev** [reply]
Could you elaborate more on your use case and how it fails?

**Reply by ganigeorgiev** [reply]
Looking at the [SQLite docs](https://www.sqlite.org/datatype3.html#type_affinity), the `NUMERIC` column type that we use for the `number` field should work fine with both `INTEGER` and `REAL` datatypes and SQLite is expected to convert the `REAL` value as `INTEGER` if it can be represented exactly as such.

If the problem is with the returned value in the Web APIs, then I guess we can normalize the value conditionally based on the `NoDecimal` option (_`OnlyInt` in v0.23.0_) and cast the float64 to int (aka. discard any fraction) but still I'm against introducing int64/bigint field type if that's what you mean because they can't be safely represented as JSON and most json encoder/decoders require some extra handling to work with such types.
